ocpp1.6/central_system.go

```diff
@@ -374,10 +383,10 @@ func (cs *centralSystem) SendRequestAsync(clientId string, request ocpp.Request,
 		return fmt.Errorf("unsupported action %v on central system, cannot send request", featureName)
 	}
 
-	cs.insertCallback(clientId, callback)
+	cs.callbackQueue.Queue(clientId, callback)
 
 	if err := cs.server.SendRequest(clientId, request); err != nil {
```

Comment:
There is a potential race condition here (consequence of my other comment):

imagine the application is invoking `SendRequestAsync` from multiple goroutines G1 and G2 at the same time (this should be supported, although I'm not doing it in the example code), there is a chance that:
1. G1 calls `cs.server.SendRequest` with request R1
2. OCPP-J layer correctly processes and enqueues R1
3. G2 calls `cs.server.SendRequest` with request R2
4. OCPP-J layer correctly processes and enqueues R2
5. G2 adds the callback to R2 via `cs.callbackQueue.Queue(...)`
6. G1 adds the callback to R1 via `cs.callbackQueue.Queue(...)`

The callback queue now holds [C2, C1], but the requests will actually be sent out like [R1, R2] -> the response from R1 will invoke C2 and so on.

There always needs to be a consistent state between enqueued requests and callbacks. 

## Suggestion

I would make the entire `SendRequestAsync` a single atomic operation with a Lock/Unlock, so it is guaranteed that enqueuing the request and setting the callback is indeed executed sequentially.

Open for other suggestions though.

Reply:
Makes sense!

Reply:
Although I'm wondering if maybe the callbacks can just be indexed by the `requestId`? And in that case perhaps callback handling belongs in the dispatcher?

Reply:
I've solved it slightly differently for now, but I think it's a nice way of encapsulating the callbacks

Reply:
The new solution also works, indeed! The new invocation doesn't feel 100% clean yet imho, but it solves the issue so I'd merge it for now (btw you really could save a few cpu cycles by appending the callback only in case the `SendRequest` succeeds, won't make a big difference though).

Reply:
That could _theoretically_ be a race condition, right? If `SendRequest` is called, the request sent and the response comes back before we have the chance to `append` it? Relatively unlikely :smiley:

Reply:
> Although I'm wondering if maybe the callbacks can just be indexed by the `requestId`? And in that case perhaps callback handling belongs in the dispatcher?

I thought a lot about passing the callbacks down to the dispatcher. It would solve the queueing issue, since the FIFO logic is handled down there already, but it would lose some transparency (`charge_point` and `central_system` wouldn't be notified of incoming responses/error anymore, since the callback would be invoked directly).

Reply:
> That could _theoretically_ be a race condition, right? If `SendRequest` is called, the request sent and the response comes back before we have the chance to `append` it? Relatively unlikely 😃

It's an async request/response. Really unlikely (unless it's a mocked synchronous test) 😅 but you're right, let's keep it as it is.